esp-hal/src/lcd_cam/lcd/dpi.rs

```diff
@@ -0,0 +1,827 @@
+//!         polarity: Polarity::IdleHigh,
+//!         phase: Phase::ShiftLow,
+//!     },
+//!     format: Format {
+//!         enable_2byte_mode: false,
+//!         ..Default::default()
+//!     },
+//!     // Send a 50x50 video
+//!     timing: FrameTiming {
+//!         horizontal_total_width: 65,
+//!         hsync_width: 5,
+//!         horizontal_blank_front_porch: 10,
+//!         horizontal_active_width: 50,
+//!
+//!         vertical_total_height: 65,
+//!         vsync_width: 5,
+//!         vertical_blank_front_porch: 10,
+//!         vertical_active_height: 50,
+//!
+//!         hsync_position: 0,
+//!     },
+//!     vsync_idle_level: Level::High,
+//!     hsync_idle_level: Level::High,
+//!     de_idle_level: Level::Low,
+//!     disable_black_region: false,
+//!     ..Default::default()
+//! };
+//!
+//! let mut dpi = Dpi::new(lcd_cam.lcd, channel.tx, 1.MHz(), config)
+//!     .with_vsync(peripherals.GPIO3)
+//!     .with_hsync(peripherals.GPIO46)
+//!     .with_de(peripherals.GPIO17)
+//!     .with_pclk(peripherals.GPIO9)
+//!     // Blue
+//!     .with_data0(peripherals.GPIO10)
+//!     .with_data1(peripherals.GPIO11)
+//!     .with_data2(peripherals.GPIO12)
+//!     .with_data3(peripherals.GPIO13)
+//!     .with_data4(peripherals.GPIO14)
+//!     // Green
+//!     .with_data5(peripherals.GPIO21)
+//!     .with_data6(peripherals.GPIO8)
+//!     .with_data7(peripherals.GPIO18)
+//!     .with_data8(peripherals.GPIO45)
+//!     .with_data9(peripherals.GPIO38)
+//!     .with_data10(peripherals.GPIO39)
+//!     // Red
+//!     .with_data11(peripherals.GPIO40)
+//!     .with_data12(peripherals.GPIO41)
+//!     .with_data13(peripherals.GPIO42)
+//!     .with_data14(peripherals.GPIO2)
+//!     .with_data15(peripherals.GPIO1);
+//!
+//! let color: u16 = 0b11111_000000_00000; // RED
+//! for chunk in dma_buf.chunks_mut(2) {
+//!     chunk.copy_from_slice(&color.to_le_bytes());
+//! }
+//!
+//! let transfer = dpi.send(false, dma_buf).map_err(|e| e.0).unwrap();
+//! transfer.wait();
+//! # }
+//! ```
+
+use core::{
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+};
+
+use fugit::HertzU32;
+
+use crate::{
+    clock::Clocks,
+    dma::{ChannelTx, DmaChannelConvert, DmaEligible, DmaError, DmaPeripheral, DmaTxBuffer, Tx},
+    gpio::{interconnect::PeripheralOutput, Level, OutputSignal},
+    lcd_cam::{
+        calculate_clkm,
+        lcd::{ClockMode, DelayMode, Lcd, Phase, Polarity},
+        BitOrder,
+        ByteOrder,
+    },
+    peripheral::{Peripheral, PeripheralRef},
+    peripherals::LCD_CAM,
+    Blocking,
+    Mode,
+};
+
+/// Represents the RGB LCD interface.
+pub struct Dpi<'d> {
+    lcd_cam: PeripheralRef<'d, LCD_CAM>,
+    tx_channel: ChannelTx<'d, Blocking, <LCD_CAM as DmaEligible>::Dma>,
+}
+
+impl<'d> Dpi<'d> {
+    /// Create a new instance of the RGB/DPI driver.
+    pub fn new<DM: Mode, CH>(
+        lcd: Lcd<'d, DM>,
+        channel: ChannelTx<'d, Blocking, CH>,
+        frequency: HertzU32,
```

Comment:
Is there max frequency that we should be checking here?

Reply:
Yes, it's checked below in `calculate_clkm`.